src/blockchain.js

```diff
@@ -1,91 +1,94 @@
-import {reduce} from './async'
+/*global ethereum*/
+import {reduce} from './async';
 
 let Blockchain = {};
 
-function isNewWeb3_1() {
-  return (typeof(web3.version) === "string");
-}
+Blockchain.Providers = {};
 
-function getAccounts(cb) {
-  if (isNewWeb3_1()) {
-    return web3.eth.getAccounts().then(function(accounts) {
-      return cb(null, accounts);
-    }).catch(function(err) {
-      return cb(err);
-    });
+Blockchain.registerProvider = function (providerName, obj) {
+  this.Providers[providerName] = obj;
```

Comment:
Assuming that this can be used as

```
EmbarkJS.Blockchain.registerProvider('someName', web3);
```
and that, internally, EmbarkJS uses `blockchainConnector.[someProviderMethod]()`, I believe this is not enough.

For example, if somebody really registers a different or custom provider object, the shape of that object may look totally different. So I think, in order to get this right, we might want to consider introducing an adapter layer that people can register as well.

e.g. Something like:

```
class MyProviderAdapter {

  constructor(provider) {
    this.provider = provider;
  }

  getAccounts() {}
  setProvider() {}
  ...
  // all the methods we rely on
}
```

EmbarkJS would then use the Adapter to create a Provider object, which can be used internally without breaking, in case the underlying provider has a different shape.

This is especially needed because we also happen to rely on utility functions such as `toWei()` which other provider might not even have.

Hope this makes sense.

Reply:
After looking at https://github.com/embark-framework/embark/pull/1150, it turns out you're kind of doing that over there.